# 🏆 The World Cup Through Data: 1930–2022
*94 Years of Football's Biggest Stage — And What to Expect in 2026*

In [1]:
import pandas as pd
import sqlite3
import os

In [17]:
conn = sqlite3.connect(':memory:')

csv_folder = '../data/raw'

for file in os.listdir(csv_folder):
    if file.endswith('.csv'):
        table_name = file.replace('.csv', '')
        df = pd.read_csv(os.path.join(csv_folder, file))
        df.to_sql(table_name, conn, if_exists='replace', index=False)
        print(f"-{table_name}: {len(df)} rows")

-groups: 159 rows
-managers: 475 rows
-qualified_teams: 625 rows
-teams: 88 rows
-player_appearances: 27432 rows
-referees: 493 rows
-referee_appearances: 1248 rows
-award_winners: 200 rows
-matches: 1248 rows
-tournaments: 30 rows
-tournament_standings: 120 rows
-substitutions: 10222 rows
-bookings: 3178 rows
-manager_appearances: 2538 rows
-players: 10401 rows
-tournament_stages: 155 rows
-confederations: 6 rows
-squads: 13843 rows
-group_standings: 626 rows
-stadiums: 240 rows
-host_countries: 31 rows
-manager_appointments: 637 rows
-referee_appointments: 668 rows
-goals: 3637 rows
-penalty_kicks: 396 rows
-awards: 8 rows
-team_appearances: 2496 rows


## Part A: Data Verification

In [3]:
tables = pd.read_sql("SELECT name FROM sqlite_master WHERE type='table'", conn)
print(f"Total tables: {len(tables)}")
tables

Total tables: 27


,name
0,groups
1,managers
2,qualified_teams
3,teams
4,player_appearances
5,referees
6,referee_appearances
7,award_winners
8,matches
9,tournaments


In [9]:
pd.read_sql("SELECT COUNT(*) AS total_goals FROM goals WHERE tournament_name = '2022 FIFA Men''s World Cup'", conn)

,total_goals
0,172


In [11]:
key_tables = ['matches', 'goals', 'bookings', 'tournaments', 'team_appearances']

for table in key_tables:
    print(f"\n{'='*60}")
    print(f"📋 {table.upper()}")
    print(f"{'='*60}")
    df = pd.read_sql(f"SELECT * FROM {table} LIMIT 3", conn)
    print(f"Columns: {list(df.columns)}\n")
    display(df)


📋 MATCHES
Columns: ['key_id', 'tournament_id', 'tournament_name', 'match_id', 'match_name', 'stage_name', 'group_name', 'group_stage', 'knockout_stage', 'replayed', 'replay', 'match_date', 'match_time', 'stadium_id', 'stadium_name', 'city_name', 'country_name', 'home_team_id', 'home_team_name', 'home_team_code', 'away_team_id', 'away_team_name', 'away_team_code', 'score', 'home_team_score', 'away_team_score', 'home_team_score_margin', 'away_team_score_margin', 'extra_time', 'penalty_shootout', 'score_penalties', 'home_team_score_penalties', 'away_team_score_penalties', 'result', 'home_team_win', 'away_team_win', 'draw']



,key_id,tournament_id,tournament_name,match_id,match_name,stage_name,group_name,group_stage,knockout_stage,replayed,...,away_team_score_margin,extra_time,penalty_shootout,score_penalties,home_team_score_penalties,away_team_score_penalties,result,home_team_win,away_team_win,draw
0,1,WC-1930,1930 FIFA Men's World Cup,M-1930-01,France vs Mexico,group stage,Group 1,1,0,0,...,-3,0,0,0-0,0,0,home team win,1,0,0
1,2,WC-1930,1930 FIFA Men's World Cup,M-1930-02,United States vs Belgium,group stage,Group 4,1,0,0,...,-3,0,0,0-0,0,0,home team win,1,0,0
2,3,WC-1930,1930 FIFA Men's World Cup,M-1930-03,Yugoslavia vs Brazil,group stage,Group 2,1,0,0,...,-1,0,0,0-0,0,0,home team win,1,0,0



📋 GOALS
Columns: ['key_id', 'goal_id', 'tournament_id', 'tournament_name', 'match_id', 'match_name', 'match_date', 'stage_name', 'group_name', 'team_id', 'team_name', 'team_code', 'home_team', 'away_team', 'player_id', 'family_name', 'given_name', 'shirt_number', 'player_team_id', 'player_team_name', 'player_team_code', 'minute_label', 'minute_regulation', 'minute_stoppage', 'match_period', 'own_goal', 'penalty']



,key_id,goal_id,tournament_id,tournament_name,match_id,match_name,match_date,stage_name,group_name,team_id,...,shirt_number,player_team_id,player_team_name,player_team_code,minute_label,minute_regulation,minute_stoppage,match_period,own_goal,penalty
0,1,G-0001,WC-1930,1930 FIFA Men's World Cup,M-1930-01,France vs Mexico,1930-07-13,group stage,Group 1,T-30,...,0,T-30,France,FRA,19',19,0,first half,0,0
1,2,G-0002,WC-1930,1930 FIFA Men's World Cup,M-1930-01,France vs Mexico,1930-07-13,group stage,Group 1,T-30,...,0,T-30,France,FRA,40',40,0,first half,0,0
2,3,G-0003,WC-1930,1930 FIFA Men's World Cup,M-1930-01,France vs Mexico,1930-07-13,group stage,Group 1,T-30,...,0,T-30,France,FRA,43',43,0,first half,0,0



📋 BOOKINGS
Columns: ['key_id', 'booking_id', 'tournament_id', 'tournament_name', 'match_id', 'match_name', 'match_date', 'stage_name', 'group_name', 'team_id', 'team_name', 'team_code', 'home_team', 'away_team', 'player_id', 'family_name', 'given_name', 'shirt_number', 'minute_label', 'minute_regulation', 'minute_stoppage', 'match_period', 'yellow_card', 'red_card', 'second_yellow_card', 'sending_off']



,key_id,booking_id,tournament_id,tournament_name,match_id,match_name,match_date,stage_name,group_name,team_id,...,given_name,shirt_number,minute_label,minute_regulation,minute_stoppage,match_period,yellow_card,red_card,second_yellow_card,sending_off
0,1,B-0001,WC-1970,1970 FIFA Men's World Cup,M-1970-01,Mexico vs Soviet Union,1970-05-31,group stage,Group 1,T-72,...,Kakhi,11,30',30,0,first half,1,0,0,0
1,2,B-0002,WC-1970,1970 FIFA Men's World Cup,M-1970-01,Mexico vs Soviet Union,1970-05-31,group stage,Group 1,T-72,...,Givi,19,31',31,0,first half,1,0,0,0
2,3,B-0003,WC-1970,1970 FIFA Men's World Cup,M-1970-01,Mexico vs Soviet Union,1970-05-31,group stage,Group 1,T-72,...,Evgeny,6,34',34,0,first half,1,0,0,0



📋 TOURNAMENTS
Columns: ['key_id', 'tournament_id', 'tournament_name', 'year', 'start_date', 'end_date', 'host_country', 'winner', 'host_won', 'count_teams', 'group_stage', 'second_group_stage', 'final_round', 'round_of_16', 'quarter_finals', 'semi_finals', 'third_place_match', 'final']



,key_id,tournament_id,tournament_name,year,start_date,end_date,host_country,winner,host_won,count_teams,group_stage,second_group_stage,final_round,round_of_16,quarter_finals,semi_finals,third_place_match,final
0,1,WC-1930,1930 FIFA Men's World Cup,1930,1930-07-13,1930-07-30,Uruguay,Uruguay,1,13,1,0,0,0,0,1,0,1
1,2,WC-1934,1934 FIFA Men's World Cup,1934,1934-05-27,1934-06-10,Italy,Italy,1,16,0,0,0,1,1,1,1,1
2,3,WC-1938,1938 FIFA Men's World Cup,1938,1938-06-04,1938-06-19,France,Italy,0,15,0,0,0,1,1,1,1,1



📋 TEAM_APPEARANCES
Columns: ['key_id', 'tournament_id', 'tournament_name', 'match_id', 'match_name', 'stage_name', 'group_name', 'group_stage', 'knockout_stage', 'replayed', 'replay', 'match_date', 'match_time', 'stadium_id', 'stadium_name', 'city_name', 'country_name', 'team_id', 'team_name', 'team_code', 'opponent_id', 'opponent_name', 'opponent_code', 'home_team', 'away_team', 'goals_for', 'goals_against', 'goal_differential', 'extra_time', 'penalty_shootout', 'penalties_for', 'penalties_against', 'result', 'win', 'lose', 'draw']



,key_id,tournament_id,tournament_name,match_id,match_name,stage_name,group_name,group_stage,knockout_stage,replayed,...,goals_against,goal_differential,extra_time,penalty_shootout,penalties_for,penalties_against,result,win,lose,draw
0,1,WC-1930,1930 FIFA Men's World Cup,M-1930-01,France vs Mexico,group stage,Group 1,1,0,0,...,1,3,0,0,0,0,win,1,0,0
1,2,WC-1930,1930 FIFA Men's World Cup,M-1930-01,France vs Mexico,group stage,Group 1,1,0,0,...,4,-3,0,0,0,0,lose,0,1,0
2,3,WC-1930,1930 FIFA Men's World Cup,M-1930-02,United States vs Belgium,group stage,Group 4,1,0,0,...,0,3,0,0,0,0,win,1,0,0


## Part B: SQL Analysis

### Chapter 1: How Has the Tournament Grown?

**B1: Number of teams, matches, and total goals per tournament**

In [20]:
tournament_grow = pd.read_sql("""
                               SELECT year, count_teams, total_matches, total_goals
                               FROM tournaments
                               LEFT JOIN (
                                   SELECT tournament_id, COUNT(*) AS total_matches
                                   FROM matches
                                   GROUP BY tournament_id
                                   ) AS match_counts
                                USING(tournament_id)
                                LEFT JOIN (
                                    SELECT tournament_id, COUNT(*) AS total_goals
                                    FROM goals
                                    GROUP BY tournament_id
                                    ) AS goal_counts
                                USING(tournament_id)
                                WHERE tournament_name LIKE '%FIFA Men''s World Cup'
                                ORDER BY year """, conn)

tournament_grow                        

,year,count_teams,total_matches,total_goals
0,1930,13,18,70
1,1934,16,17,70
2,1938,15,18,84
3,1950,13,22,88
4,1954,16,26,140
5,1958,16,35,126
6,1962,16,32,89
7,1966,16,32,89
8,1970,16,32,95
9,1974,16,38,97


The first early World Cup era (1930–1978) stayed small around 13–16 teams at the beginning playing only 18 matches. As the years passed, the modern football era took place, the participating teams kept growing up to 32 teams in the year of 2022 with 64 matches played. Regarding the total goals scored, we cannot skip the 1954 World Cup with the remarkable 140 goals scored in only 26 matches — the highest goals-per-match rate in World Cup history. Similar total goals jumps can also be noticed in later years: 1982 with 146 goals and 1998 with 171 goals, both coinciding with tournament expansions (16→24 teams and 24→32 teams respectively).

**B2: Unique countries by decade — how has the World Cup's global reach expanded?**

In [33]:
unique_countries_total = pd.read_sql("""
                                        SELECT COUNT(DISTINCT team_name) AS unique_countries
                                        FROM team_appearances
                                        WHERE tournament_name LIKE '%FIFA Men''s World Cup' """, conn)
unique_countries_total

,unique_countries
0,85


In [34]:
unique_countries_decade = pd.read_sql("""
                                        SELECT SUBSTR(tournament_name, 1, 3) || '0s' AS decade,
                                        COUNT(DISTINCT team_name) AS unique_countries
                                        FROM team_appearances
                                        WHERE tournament_name LIKE '%FIFA Men''s World Cup'
                                        GROUP BY decade
                                        ORDER BY decade ASC """, conn)
unique_countries_decade

,decade,unique_countries
0,1930s,27
1,1950s,26
2,1960s,19
3,1970s,32
4,1980s,34
5,1990s,45
6,2000s,44
7,2010s,50
8,2020s,32


A total of 85 unique countries have participated in the men's World Cup across its 92-year history. The 1940s are absent from the data due to World War II — no tournaments were held in 1942 or 1946. The biggest jumps in global reach came in the 1970s (32 countries) and 1990s (45 countries), reflecting both tournament expansions and the growing worldwide popularity of football. The 2020s show only 32 countries, but this represents a single tournament (Qatar 2022) rather than a decline.

**B3: Host country performance — how far did each host nation go in their own World Cup?**

In [30]:
pd.read_sql("SELECT * FROM host_countries LIMIT 5", conn)

,key_id,tournament_id,tournament_name,team_id,team_name,team_code,performance
0,1,WC-1930,1930 FIFA Men's World Cup,T-84,Uruguay,URY,champions
1,2,WC-1934,1934 FIFA Men's World Cup,T-41,Italy,ITA,champions
2,3,WC-1938,1938 FIFA Men's World Cup,T-30,France,FRA,quater-finals
3,4,WC-1950,1950 FIFA Men's World Cup,T-09,Brazil,BRA,runners-up
4,5,WC-1954,1954 FIFA Men's World Cup,T-75,Switzerland,CHE,quarter-final


In [31]:
pd.read_sql("SELECT * FROM qualified_teams LIMIT 5", conn)

,key_id,tournament_id,tournament_name,team_id,team_name,team_code,count_matches,performance
0,1,WC-1930,1930 FIFA Men's World Cup,T-03,Argentina,ARG,5,final
1,2,WC-1930,1930 FIFA Men's World Cup,T-06,Belgium,BEL,2,group stage
2,3,WC-1930,1930 FIFA Men's World Cup,T-07,Bolivia,BOL,2,group stage
3,4,WC-1930,1930 FIFA Men's World Cup,T-09,Brazil,BRA,2,group stage
4,5,WC-1930,1930 FIFA Men's World Cup,T-13,Chile,CHL,3,group stage


In [32]:
pd.read_sql("SELECT * FROM host_countries", conn)

,key_id,tournament_id,tournament_name,team_id,team_name,team_code,performance
0,1,WC-1930,1930 FIFA Men's World Cup,T-84,Uruguay,URY,champions
1,2,WC-1934,1934 FIFA Men's World Cup,T-41,Italy,ITA,champions
2,3,WC-1938,1938 FIFA Men's World Cup,T-30,France,FRA,quater-finals
3,4,WC-1950,1950 FIFA Men's World Cup,T-09,Brazil,BRA,runners-up
4,5,WC-1954,1954 FIFA Men's World Cup,T-75,Switzerland,CHE,quarter-final
5,6,WC-1958,1958 FIFA Men's World Cup,T-74,Sweden,SWE,runners-up
6,7,WC-1962,1962 FIFA Men's World Cup,T-13,Chile,CHL,third place
7,8,WC-1966,1966 FIFA Men's World Cup,T-28,England,ENG,champions
8,9,WC-1970,1970 FIFA Men's World Cup,T-46,Mexico,MEX,quater-finals
9,10,WC-1974,1974 FIFA Men's World Cup,T-86,West Germany,DEU,champions


In [43]:
host_performance = pd.read_sql("""
                                SELECT
                                    CASE 
                                        WHEN performance LIKE '%quater%' OR performance LIKE '%quarter%' THEN 'quarter-finals'
                                        ELSE performance
                                    END AS performance_clean,
                                    COUNT (*) AS total
                                FROM host_countries
                                WHERE tournament_name LIKE '%FIFA Men''s World Cup'
                                GROUP BY performance_clean """, conn)

host_performance

,performance_clean,total
0,champions,6
1,fourth place,2
2,group stage,2
3,quarter-finals,5
4,round of 16,2
5,runners-up,2
6,second group stage,1
7,third place,3


Host nations have a remarkable track record — 6 out of 22 hosts became champions, with another 2 finishing as runners-up and 3 taking third place. That means half of all hosts reached at least the semi-finals. Only 2 hosts (South Africa 2010 and Qatar 2022) were eliminated in the group stage. With three co-hosts in 2026 (USA, Mexico, and Canada), it will be interesting to see whether the historical host advantage translates when the tournament is shared across countries.

### Chapter 2: How Has Scoring Changed?

**B4: Average goals per match per tournament — the signature curve of football analytics**

In [44]:
tournament_grow['goals_per_match'] = round(tournament_grow['total_goals'] / tournament_grow['total_matches'], 2)
tournament_grow

,year,count_teams,total_matches,total_goals,goals_per_match
0,1930,13,18,70,3.89
1,1934,16,17,70,4.12
2,1938,15,18,84,4.67
3,1950,13,22,88,4.00
4,1954,16,26,140,5.38
5,1958,16,35,126,3.60
6,1962,16,32,89,2.78
7,1966,16,32,89,2.78
8,1970,16,32,95,2.97
9,1974,16,38,97,2.55


The 1954 World Cup stands as both the peak and the outlier at 5.38 goals per match — nearly double any modern tournament. From that point on, the average settled around 2.5 goals per match, reflecting a shift toward more tactical, defensively organized football. The modern 32-team era (1998–2022) shows remarkable stability, hovering between 2.23 and 2.69 goals per match despite nearly doubling the number of games — a clear sign that the style of play, not the volume of matches, drives the average.

**B5: Goal breakdown by type — are modern World Cups more penalty-dependent?**

In [49]:
goal_breakdown = pd.read_sql("""
                                SELECT SUBSTR(tournament_name, 1, 4) AS year, SUM(penalty) AS penalty_goals, SUM(own_goal) AS own_goals, 
                                COUNT(*) - SUM(penalty) - SUM(own_goal) AS open_play_goals
                                FROM goals
                                WHERE tournament_name LIKE '%FIFA Men''s World Cup'
                                GROUP BY year
                                ORDER BY year
                                """, conn)
goal_breakdown

,year,penalty_goals,own_goals,open_play_goals
0,1930,1,1,68
1,1934,3,0,67
2,1938,3,2,79
3,1950,3,0,85
4,1954,7,4,129
5,1958,7,0,119
6,1962,8,0,81
7,1966,8,2,79
8,1970,5,1,89
9,1974,6,3,88


In [50]:
goal_breakdown['penalty_pct'] = round(goal_breakdown['penalty_goals'] * 100 / (goal_breakdown['penalty_goals'] 
                                + goal_breakdown['own_goals'] + goal_breakdown['open_play_goals']), 1)
goal_breakdown

,year,penalty_goals,own_goals,open_play_goals,penalty_pct
0,1930,1,1,68,1.4
1,1934,3,0,67,4.3
2,1938,3,2,79,3.6
3,1950,3,0,85,3.4
4,1954,7,4,129,5.0
5,1958,7,0,119,5.6
6,1962,8,0,81,9.0
7,1966,8,2,79,9.0
8,1970,5,1,89,5.3
9,1974,6,3,88,6.2


Penalty goals have grown in raw numbers over the decades, but much of that increase is explained by more matches being played — the percentage has hovered between 5–11% for most tournaments. The clear outlier is 2018 (13%), which coincided with the introduction of VAR technology, awarding penalties that referees might previously have missed. That same year also saw a record 12 own goals, though this is likely coincidental rather than VAR-related.

**B6: Goal timing distribution — when do World Cup goals happen?**

In [55]:
goal_timing = pd.read_sql("""
                            SELECT 
                                CASE 
                                    WHEN minute_regulation <= 15 THEN '1-15'
                                    WHEN minute_regulation <= 30 THEN '16-30'
                                    WHEN minute_regulation <= 45 THEN '31-45'
                                    WHEN minute_regulation <= 60 THEN '46-60'
                                    WHEN minute_regulation <= 75 THEN '61-75'
                                    WHEN minute_regulation <= 90 THEN '76-90'
                                    WHEN minute_regulation > 90 THEN '90+'
                                END AS minute_bucket,
                                COUNT (*) AS total_goals
                            FROM goals
                            WHERE tournament_name LIKE '%FIFA Men''s World Cup'
                            GROUP BY minute_bucket
                            """, conn)
goal_timing

,minute_bucket,total_goals
0,1-15,349
1,16-30,389
2,31-45,407
3,46-60,442
4,61-75,483
5,76-90,579
6,90+,71


Goals steadily increase as the match progresses, with the 76–90 minute period being the most dangerous (579 goals — nearly double the 1–15 minute bucket). Overall, second halves produce roughly 30% more goals than first halves, reflecting tired defenders, tactical substitutions, and the urgency of chasing a result late in the game.

**B7: Blowout analysis — do expanded tournaments produce more lopsided matches?**

In [60]:
blowout_analysis = pd.read_sql("""
                                SELECT SUBSTR(tournament_name, 1, 4) AS year,
                                SUM(CASE WHEN ABS(home_team_score - away_team_score) >= 3 THEN 1 ELSE 0 END) AS blowouts,
                                COUNT(*) AS total_matches
                                FROM matches
                                WHERE tournament_name LIKE '%FIFA Men''s World Cup'
                                GROUP BY year
                                ORDER BY year
                                """, conn)
blowout_analysis['blowout_pct'] = round(blowout_analysis['blowouts'] * 100.0 / blowout_analysis['total_matches'], 1)
blowout_analysis

,year,blowouts,total_matches,blowout_pct
0,1930,10,18,55.6
1,1934,2,17,11.8
2,1938,4,18,22.2
3,1950,7,22,31.8
4,1954,11,26,42.3
5,1958,9,35,25.7
6,1962,3,32,9.4
7,1966,4,32,12.5
8,1970,8,32,25.0
9,1974,8,38,21.1


Expanded tournaments did not produce more blowouts — the opposite, in fact. The early era was far more lopsided, with 1930 (55.6%) and 1954 (42.3%) leading by a wide margin. The modern 32-team era has stabilized between 10–19%, and the 1998 expansion from 24 to 32 teams barely changed the blowout rate. This suggests that the qualifying process has kept the overall standard competitive, even as more teams enter the tournament.